src/vs/workbench/contrib/notebook/browser/view/cellPart.ts

```diff
@@ -57,49 +64,149 @@ export abstract class CellPart extends Disposable {
 	updateForExecutionState(element: ICellViewModel, e: ICellExecutionStateChangedEvent): void { }
 }
 
+export abstract class CellOverlayPart extends Disposable {
+	protected currentCell: ICellViewModel | undefined;
+	protected cellDisposables = new DisposableStore();
+
+	constructor() {
+		super();
+	}
+
+	/**
+	 * Prepare model for cell part rendering
+	 * No DOM operations recommended within this operation
+	 */
+	prepareRenderCell(element: ICellViewModel): void { }
+
+	/**
+	 * Update the DOM for the cell `element`
+	 */
+	renderCell(element: ICellViewModel): void {
+		this.currentCell = element;
+		this.didRenderCell(element);
+	}
+
+	protected didRenderCell(element: ICellViewModel): void { }
+
+	/**
+	 * Dispose any disposables generated from `didRenderCell`
+	 */
+	unrenderCell(element: ICellViewModel): void {
+		this.currentCell = undefined;
+		this.cellDisposables.clear();
+	}
+
+	/**
+	 * Update internal DOM (top positions) per cell layout info change
+	 * Note that a cell part doesn't need to call `DOM.scheduleNextFrame`,
+	 * the list view will ensure that layout call is invoked in the right frame
+	 */
+	updateInternalLayoutNow(element: ICellViewModel): void { }
+
+	/**
+	 * Update per cell state change
+	 */
+	updateState(element: ICellViewModel, e: CellViewModelStateChangeEvent): void { }
+
+	/**
+	 * Update per execution state change.
+	 */
+	updateForExecutionState(element: ICellViewModel, e: ICellExecutionStateChangedEvent): void { }
+}
+
 export class CellPartsCollection {
+	private _scheduledOverlayRendering: IDisposable | undefined;
+	private _scheduledOverlayUpdateState: IDisposable | undefined;
+	private _scheduledOverlayUpdateExecutionState: IDisposable | undefined;
 
 	constructor(
-		private readonly parts: readonly CellPart[],
+		private readonly contentParts: readonly CellContentPart[],
+		private readonly overlayParts: readonly CellOverlayPart[]
 	) { }
 
-	concat(other: readonly CellPart[]): CellPartsCollection {
-		return new CellPartsCollection(this.parts.concat(other));
+	concatContentPart(other: readonly CellContentPart[]): CellPartsCollection {
+		return new CellPartsCollection(this.contentParts.concat(other), this.overlayParts);
 	}
 
-	renderCell(element: ICellViewModel): void {
-		for (const part of this.parts) {
+	concatOverlayPart(other: readonly CellOverlayPart[]): CellPartsCollection {
+		return new CellPartsCollection(this.contentParts, this.overlayParts.concat(other));
+	}
+
+	scheduleRenderCell(element: ICellViewModel): void {
+		// prepare model
+		for (const part of this.contentParts) {
+			part.prepareRenderCell(element);
+		}
+
+		for (const part of this.overlayParts) {
+			part.prepareRenderCell(element);
+		}
+
+		// render content parts
+		for (const part of this.contentParts) {
 			part.renderCell(element);
 		}
+
+		// schedule overlay parts rendering
+		this._scheduledOverlayRendering?.dispose();
+
+		this._scheduledOverlayRendering = DOM.modify(() => {
```

Comment:
Does this scheduled rendering need to be keyed on the `element` passed in? Otherwise if `scheduleRenderCell` is called twice with two different `element` values, won't the first one never be rendered?

Reply:
A part can only be rendered against one cell at a time, and before it's being rendered against a new cell, it would be unrendered against previous cell thus the single scheduled job is what we want.